L13-python-interceptors.md

```diff
@@ -0,0 +1,274 @@
+
+## Rationale
+
+### The Benefit of Introducing "Continuation"
+
+Quote from mehrdada@ [design doc](https://github.com/grpc/proposal/blob/361846400834164d21c12ca6cf60e8cbf501d6cd/L13-Python-Interceptors.md):
+> All of the methods are passed a callable value `Continuation`, whose
+> return type matches the return type of that is expected of the
+> function it is passed to. `Continuation` is the continuation that
+> progresses the RPC forward by calling the next interceptor
+> in the chain, or the core RPC Continuation, in case of the last
+> interceptor.  It is the responsibility of the interceptor
+> to call `Continuation` should it chooses to continue with the RPC,
+> but is not required to do so.  The ability to invoke the
+> `Continuation` not exactly once is the reason this design is
+> advantageous over the system automatically invoking
+> the continuation upon returning from interceptor.
+
+### History of Its Implementation
+
+The Python interceptors were requested by joeblubaugh@ from Twitter in [grpc/grpc#8767](https://github.com/grpc/grpc/issues/8767) in 2016 November. According to the discussion, the Python teams decided that this feature should be done by the community, since the `grpc.Channel` interface can be easily re-implemented. 
+
+Fortunately, rnburn@ was willing to implement this feature to push the progress of [OpenTracing](https://opentracing.io/), and submitted a pull request [grpc/grpc#10256](https://github.com/grpc/grpc/pull/10256) in 2017 March. The pull request defines the majority of the semantics of today's gRPC Python interceptors implementation, which wraps over the existing abstraction of `grpc.Channel` and `grpc.Server`.
+
+However, rnburn@ realized (see [comment](https://github.com/grpc/grpc/pull/10256#issuecomment-311495482)) that this approach prevents the capability for gRPC Python to provide same interceptor granularity as other implementation (Java, Node.js), and halt the work on this PR. To provide the most effective interceptor semantics, the implementation has to wrap over low-level C-Core API like `startBatch`, which sends batches of gRPC operations into C-Core. The potential amount of work is much more considerable than expected, it might require to refactor abstractions of gRPC or have hacky injections of logic.
+
+To balance the cost and benefit, the Python team decided to carry on with the approach of wrapping over high-level abstraction. Then, merhdada@ wrote a gRFC ([grpc/proposal#39](https://github.com/grpc/proposal/pull/39)) that extends the functionality of interceptor along with two PR ([grpc/grpc#12778](https://github.com/grpc/grpc/pull/12778) and [grpc/grpc#13722](https://github.com/grpc/grpc/pull/13722)) based on rnburn@'s implementation. However, the PR itself conflicts with the gRFC, more specifically, the client-side interceptor API is changed, and the server-side interceptor semantics are not fully implemented.
```

Comment:
The Python team decided to carry on with rnburn@'s implementation that wrapped the high-level Server and Channel abstractions. Then, merhdada@ wrote a draft gRFC ([grpc/proposal#39](https://github.com/grpc/proposal/pull/39)) that defined the functionality of interceptor along with two PRs ([grpc/grpc#12778](https://github.com/grpc/grpc/pull/12778) and [grpc/grpc#13722](https://github.com/grpc/grpc/pull/13722)) based on rnburn@'s implementation. However, the implementation itself conflicts with the original draft gRFC, leading to the current gRFC, which describes the interceptor APIs as actually implemented in the gRPC Python library.